Updating documentation around assets referencing other files (#1882)
o3de/o3de.org: content/docs/user-guide/assets/pipeline/asset-builders.md

---
linkTitle: Asset Builders 
title: Asset Builders
description: Asset Builders use configurable rules to generate product assets from jobs that are dispatched by Asset Processor in Open 3D Engine (O3DE).
weight: 400
toc: true
---

An **Asset Builder** is a bundle of code that **Asset Processor** runs to generate product assets. Asset Builders can be distributed as Gems and might use third-party libraries to process assets. The **Scene Processing** Gem, for example, contains the Asset Builders that process `.fbx` files, and uses the [Open Asset Import Library](https://github.com/assimp/assimp) to parse `.fbx` files.

{{< note >}}
The Open Asset Import Library supports [many scene formats](https://github.com/assimp/assimp/blob/master/doc/Fileformats.md). You can experiment with additional formats by editing `o3de/Registry/sceneassetimporter.setreg` and adding format extensions to the `"SupportedFileTypeExtensions"` list.
{{< /note >}}

## Anatomy of an asset builder

Asset Builders have three core components: a **Descriptor** for the builder, and handlers for **Create Jobs** and **Process Job** requests.

> ![Simple Job](/images/user-guide/assets/pipeline/asset_builders/asset_builder_basics.png)

Reviewer: We can embed SVG versions of these files. If you replace the PNGs with SVGs, I can create an issue to revisit this doc and make some adjustments so they scale nicely.

Author: I don't have SVG versions of those images

| debug              | Debug mode for the create and process job of the specified file. <br/>Debug mode optionally uses `-input`, `-output`, `-module`, `-port` and `-gameroot`. For example: `-debug Objects\Tutorials\shapes.fbx`. |                                                                                                                                  
| debug_create       | Debug mode for the create job of the specified file.                                                                                                                                                          |
| debug_process      | Debug mode for the process job of the specified file.                                                                                                                                                         |
| tags               | Additional tags to add to the debug platform for job processing. One tag can be supplied per option.                                                                                                          |
| platform           | Platform to use for debugging, for example, `pc` |

## Loading other files

> Sometimes, when authoring a builder, it may be necessary to load other files besides the primary source file to finish a given process. Here are some scenarios that this can come up, and best practices for handling them.

Reviewer: The second sentence seems out of place with the new organization.

```suggestion
Sometimes, when authoring a builder, it may be necessary to load other files besides the primary source file to finish a given process.
```

### Understanding file locations

![File Locations](/images/user-guide/assets/pipeline/asset_builders/asset_builder_file_locations.png)

Files relevant to asset processing can exist in these locations:
1. The Asset Cache is where all product assets exist.
1. Scan directories are the only place source assets can exist.
1. Source files (not assets) can be on any drive in any location.

> Scan directories are mostly these locations. Scan directories can be elsewhere, but these are the primary ones. Read more about [scan directories here:](/docs/user-guide/assets/pipeline/scan-directories/)

Reviewer: A suggestion for clarity.

```suggestion
[Scan directories](/docs/user-guide/assets/pipeline/scan-directories/) are monitored by Asset Processor for new and updated source assets. Scan directories can be user defined, however, the primary scan directories are in the following locations:
```

1. The current project's directory.
1. The `Assets` folder for each Gem enabled for the project.

> Individual Gems, the O3DE project in use, and the engine itself can all be installed to different drives.

Reviewer: "In use" strikes me as a bit odd, a suggestion.

```suggestion
Individual Gems, the active O3DE project, and the engine itself can all be installed to different drives.
```

#### Types of files in asset processing

![All files in asset processing](/images/user-guide/assets/pipeline/asset_builders/asset_builder_all_dependencies.png)

* Source assets

>    * Source assets are files that exist in scan directories for an O3DE project, that match an Asset Builder's registered pattern, and result in a job being created when create jobs is called with the source asset for the Asset Builder.

Reviewer: Suggest consistently capitalizing Create Jobs.

```suggestion
   * Source assets are files that exist in scan directories for an O3DE project, that match an Asset Builder's registered pattern, and result in a job being created when Create Jobs is called with the source asset for the Asset Builder.
```

   * Read more about [source assets here.](/docs/user-guide/assets/pipeline/source-assets/)
* Non-asset source files
   * Non-asset source files are files that may or may not be in a scan directory, that are loaded and referenced when processing a source asset.
   * The difference between a source asset and a non-asset source file is either:
      * The file is not in a scan directory,
      * The file does not match a file pattern for any asset builder descriptor, or
      * The file does not generate a job when create jobs is called for the file for matching asset builders
* Product assets

>    * Product assets are the run-time ready output of an asset processing job.

Reviewer: Minor, we're trying to use runtime consistently rather than run-time.

```suggestion
   * Product assets are the runtime ready output of an asset processing job.
```

> * Intermediate assets

Reviewer: Feels like intermediate assets should precede product assets in this list.

   * Intermediate assets are source assets that are generated as a product of an asset processing job.
   * Read more about [intermediate assets here.](/docs/user-guide/assets/pipeline/intermediate-assets/)

> #### References from source assets and non-asset source files to other files

Reviewer: Suggest making this an H3 so that it appears in the page ToC on the right.

```suggestion
### References from source assets and non-asset source files to other files
```

![Source asset references](/images/user-guide/assets/pipeline/asset_builders/source_asset_references.png)

> Source assets and non-asset source files can reference each other in many different ways, and each case may require unique handling at the time of authoring an asset builder.

Reviewer: minor

```suggestion
Source assets and non-asset source files can reference each other in many different ways, and each case might require unique handling at the time of authoring an asset builder.
```

Resolving paths to find the file on disk is often done within the asset builder. When dealing with path references from the source asset to other files, keep these things in mind:
* The referenced asset may not be on the same drive as the asset with the reference. This includes scan directories, which can be on different drives. This means relative paths may not be resolvable.
* Some operating systems have case sensitive paths, some do not. When resolving paths in an asset builder, it's important to do so in a way that will be stable across operating systems.
* File locations relative to scan directories, and in the asset cache will be stable for all members of a team, but scan directories themselves, and files outside scan directories may not have the same paths for all team members.
* Path resolving logic should be consistent, and predictable. Content creators will want to understand how paths are resolved when processing assets, so they can better manage their content.

> #### References from product assets to other files

Reviewer: Suggest making this an H3 so it shows up in the page ToC.

```suggestion
### References from product assets to other files
```

Source assets and non-asset source files may have these references:
* source assets by UUID
* source assets by a path
* non-asset source files in a scan directory by a path
* non-asset source files outside a scan directory by a path
* product assets by asset ID
* product assets by a path

Path references from source asset and non-asset source files may be defined in these ways:
* Relative to a scan directory
* Relative to the asset with the reference
* Absolute
* Relative to a directory that may be difficult to infer from the contents or location of the source asset

Source assets come in two forms, relative to how these references work:
* A format that the builder author can modify and control.
  * Example: A custom format defined for O3DE, like prefab files.
* A format that the builder author has no control over.
  * Example: A format defined outside of O3DE, like FBX files.

When authoring a builder, if the format of the source asset is something the author has control over, then it's useful to manage how the source asset references other files.

> * When referencing a source asset or product asset, it's preferred to reference by UUID or Asset ID, because these will be more stable than a path. Read more about [asset identifiers here.](/docs/user-guide/assets/pipeline/asset-dependencies-and-identifiers/#asset-identifiers)

Reviewer 2: Nit: Maybe we should force people to use the UUID/Asset ID to reference source/product asset. The APIs for resolving file path to asset ID are for the compatibility with the legacy cry system and should be avoided in any new O3DE systems.

Author: We can't force it because many source assets are going to be in formats we don't control. For example, an FBX file can have paths to textures baked in.

* The referenced asset may not be on the same drive as the asset with the reference. This includes scan directories, which can be on different drives. This means relative paths may not be resolvable.
* Some operating systems have case sensitive paths, some do not. When resolving paths in an asset builder, it's important to do so in a way that will be stable across operating systems.
* File locations relative to scan directories, and in the asset cache will be stable for all members of a team, but scan directories themselves, and files outside scan directories may not have the same paths for all team members.
* Path resolving logic should be consistent, and predictable. Content creators will want to understand how paths are resolved when processing assets, so they can better manage their content.

#### References from product assets to other files

![Product asset references](/images/user-guide/assets/pipeline/asset_builders/product_asset_references.png)

Product assets should only have these references:
* Product assets by asset ID

> * Product assets by a relative path from the asset with the reference

Reviewer 2: Maybe we can add a note here to avoid using relative path if possible.

Author: Added a note

{{< /note >}}

### Process job

Process Job generates the product asset and product dependencies. The Asset Builder receives a `ProcessJobRequest` from Asset Processor containing info on the source asset to process. The Asset Builder responds with a `ProcessJobResponse`. The function of `ProcessJobResponse` is to process the source asset and return information about the product assets it creates, including sub IDs and product dependencies.

`ProcessJobResponse` places the product assets in a temp directory. Asset Processor moves the product assets to the **Asset Cache** and populates the **Asset Database** and **Asset Catalog** with information it uses to track the product assets, product dependencies, and the jobs that produced them.

Process Job is multithreaded. Several Asset Builders can run multiple process jobs simultaneously.

## Loading other files

Sometimes, when authoring a builder, it may be necessary to load other files besides the primary source file to finish this process. Here are some scenarios that this can come up, and best practices for handling them.

### Understanding file locations

> Files relevant to asset processing can exist in three core locations:

Author: I've cleaned up the text here.

1. The Asset Cache is where all product assets exist.
1. Scan directories are the only place source assets can exist.

> 1. Source files (not assets) can be on any drive in any location.

Author: Source assets have to be in scan directories, but source files can be anywhere

> 1. Source files (not assets) can be on any drive in any location.

Author: I've added more images to try and show the difference.

Scan directories tend to be one of two core locations. There are other locations, but these are the primary ones. Read more about [scan directories here:](/docs/user-guide/assets/pipeline/scan-directories/)

> 1. The root folder for the current project.

Author: I've reworded this.

1. The `Assets` folder for each Gem enabled for the project.

Individual Gems, the O3DE project in use, and the engine itself can all be installed to different drives.

#### Types of files in asset processing

* Source assets

>    * Source assets are files that exist in scan directories for an O3DE project, that match an Asset Builder's registered pattern, and result in a job being created when create jobs is called with the source asset for the Asset Builder.

Author: One of the things I'm trying to explain is the difference between source assets and source files. Source files may also include data to be processed into the product asset content. Source files don't have to be in a scan directory, but source assets do.

>    * Source assets are files that exist in scan directories for an O3DE project, that match an Asset Builder's registered pattern, and result in a job being created when create jobs is called with the source asset for the Asset Builder.

Author: I've added some images to show where files can be.

   * Read more about [source assets here.](/docs/user-guide/assets/pipeline/source-assets/)

> * Non-asset source files

Author: This guide is for asset builders, and I figured most asset builder authors would be thinking the source assets and files before the product assets. The builders they are writing are transforming the source assets and files into the product assets.

`ProcessJobResponse` places the product assets in a temp directory. Asset Processor moves the product assets to the **Asset Cache** and populates the **Asset Database** and **Asset Catalog** with information it uses to track the product assets, product dependencies, and the jobs that produced them.

Process Job is multithreaded. Several Asset Builders can run multiple process jobs simultaneously.

> 
> ## Asset Builder Tips and Tricks
> 
> ### Loading Other Files

Reviewer: Having these two headings back-to-back looks a bit odd on page. Suggest removing the Tips and Tricks H2 and moving all subsequent headings up a level. Also suggesting sentence casing for headings.

```suggestion
## Loading other files
```

Sometimes, when authoring a builder, it may be necessary to load other files besides the primary source file to finish this process. Here are some scenarios that this can come up, and best practices for handling them.

#### Understanding File Locations for Asset Processing

Files relevant to asset processing can exist in three core locations:
1. The Asset Cache is where all product assets exist.
1. Scan directories are the only place source assets can exist.
1. Source files (not assets) can be on any drive in any location.

Scan directories tend to be one of two core locations. There are other locations, but these are the primary ones. Read more about [scan directories here:](/docs/user-guide/assets/pipeline/scan-directories/)
1. The asset folder for the current project.
1. The asset folder for each gem enabled for the project.

> 

Author: Not that I know of, I've added a ticket to my Document Asset Pipeline Systems Epic to cover this.

Individual Gems, the O3DE project in use, and the engine itself can all be installed to different drives.

If you have source assets referencing each other, the safest way to reference one source asset from another is with that asset's UUID, because it will be stable across drives. Read more about [asset identifiers here.](/docs/user-guide/assets/pipeline/asset-dependencies-and-identifiers/#asset-identifiers) Relative paths between source assets may not function as expected, if those source assets are in scan folders across different root directories, there may not be a way to build that reference. Also, different members of a team may install their project, engine, and gems differently than each other, making relative paths in source assets not stable for the team.

> The asset cache is more stable, relative paths between two product assets won't change unless the source asset that generating them is moved, or the associated builder is updated to change how it outputs product assets. Even still, it's recommended to have product assets reference each other with asset ID and not relative path, because asset IDs are more stable long term than relative paths.

Author: I've just done a big re-write of this section.

> The asset cache is more stable, relative paths between two product assets won't change unless the source asset that generating them is moved, or the associated builder is updated to change how it outputs product assets. Even still, it's recommended to have product assets reference each other with asset ID and not relative path, because asset IDs are more stable long term than relative paths.

Author: This section was re-written and this text is gone

Product assets should never reference other product assets with absolute path because this will result in the hash of the contents of product assets being unique per machine that generates product assets. When authoring a builder, the complete lifecycle of asset management for a project should be kept in mind. Keeping product assets stable across machines will ensure that the gathering modified assets step of generating a patch for a live game is accurate. You can read more about [asset bundling here.](docs/user-guide/packaging/asset-bundler/overview/)

> #### Loading Product Assets in Process Job

Reviewer: As per previous heading suggestion. Only shortened 3 characters, but every little bit helps. If you can come up with a shorter heading, please do.

```suggestion
### Load product assets in Process Job
```

If the processing of one job requires loading the output product asset of another job, then a job dependency should be declared against that job, and when possible the specific product should be defined in this dependency.

> When setting up asset references, be aware that the source asset with the outgoing reference to a product asset may be on two different drive roots, so relative paths may not be possible between the source asset and the referenced product asset.

Author: I've updated it with that suggestion

Read more about [job dependencies here.](/docs/user-guide/assets/pipeline/asset-dependencies-and-identifiers/#job-dependencies)

> #### Loading Product Assets in Create Jobs

Reviewer: ```suggestion
### Load product assets in Create Jobs
```

If that isn't possible, then intermediate assets can provide a path to manage this processing order. To do this, the initial job should be updated to output an intermediate asset to be used as the source asset for the job to do the real work. The new job creating this intermediate asset should emit the product asset to be loaded in the middle job's create jobs function as a product specific job dependency.

Read more about [intermediate assets here.](/docs/user-guide/assets/pipeline/intermediate-assets/) Read more about [job dependencies here.](/docs/user-guide/assets/pipeline/asset-dependencies-and-identifiers/#job-dependencies)

This is an uncommon situation, so we're looking for use cases to help guide improvements to be made in this area. If you have a need to load a product asset during the create jobs step of asset processing, please create [a ticket](https://github.com/o3de/o3de/issues) or reach out in the [O3DE Discord](https://{{< links/o3de-discord >}})

> #### Loading Other Source Assets or non-asset files

Reviewer: ```suggestion
### Load other source assets or non-asset files
```
